pkg/controller/nodelink/nodelink_controller.go

```diff
@@ -0,0 +1,488 @@
+
+	if modNode.Labels == nil {
+		modNode.Labels = map[string]string{}
+	}
+
+	for k, v := range machine.Spec.Labels {
+		klog.V(4).Infof("Copying label %s = %s", k, v)
+		modNode.Labels[k] = v
+	}
+
+	addTaintsToNode(modNode, machine)
+
+	if !reflect.DeepEqual(node, modNode) {
+		klog.V(3).Infof("Node %q has changed, updating", modNode.GetName())
+		if err := r.client.Update(context.Background(), modNode); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error updating node: %v", err)
+		}
+	}
+
+	// TODO(alberto): bring nodeRef logic here and drop cluster-api node controller
+	return reconcile.Result{}, nil
+}
+
+// nodeRequestFromMachine returns a reconcile.request for the node backed by the received machine
+func (r *ReconcileNodeLink) nodeRequestFromMachine(o handler.MapObject) []reconcile.Request {
+	klog.V(3).Infof("Watched machine event, finding node to reconcile.Request")
+	// get machine
+	machine := &mapiv1beta1.Machine{}
+	if err := r.client.Get(
+		context.Background(),
+		client.ObjectKey{
+			Namespace: o.Meta.GetNamespace(),
+			Name:      o.Meta.GetName(),
+		},
+		machine,
+	); err != nil {
+		klog.Errorf("No-op: Unable to retrieve machine %s/%s from store: %v", o.Meta.GetNamespace(), o.Meta.GetName(), err)
+		return []reconcile.Request{}
+	}
+
+	if machine.DeletionTimestamp != nil {
+		klog.V(3).Infof("No-op: Machine %q has a deletion timestamp", o.Meta.GetName())
+		return []reconcile.Request{}
+	}
+
+	// find node
+	node, err := r.findNodeFromMachine(machine)
+	if err != nil {
+		klog.Errorf("No-op: Failed to find node for machine %q: %v", machine.GetName(), err)
+		return []reconcile.Request{}
+	}
+	if node != nil {
+		return []reconcile.Request{
+			{
+				NamespacedName: client.ObjectKey{
+					Namespace: node.GetNamespace(),
+					Name:      node.GetName(),
+				},
+			},
+		}
+	}
+
+	klog.V(3).Infof("No-op: Node for machine %q not found", machine.GetName())
+	return []reconcile.Request{}
+}
+
+// findNodeFromMachine tries to a node from by providerIP and fallback to find by IP
+func (r *ReconcileNodeLink) findNodeFromMachine(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
+		klog.V(3).Infof("Finding node from machine nil, returning...")
+		return nil, nil
+	}
+
+	klog.V(3).Infof("Finding node from machine %q", machine.GetName())
+	node, err := r.findNodeFromMachineByProviderID(machine)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by ProviderID: %v", machine.GetName(), err)
+	}
+	if node != nil {
+		return node, nil
+	}
+
+	node, err = r.findNodeFromMachineByIP(machine)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find node from machine %q by internal IP: %v", machine.GetName(), err)
+	}
+	return node, nil
+}
+
+func (r *ReconcileNodeLink) findNodeFromMachineByProviderID(machine *mapiv1beta1.Machine) (*corev1.Node, error) {
+	if machine == nil {
```

Comment:
Should this be an error case?

Reply:
I think we should avoid these checks. It masks other problems. We shouldn't be trying to find a node  on a machine that we don't have (i.e., machine is `nil`).

Reply:
If this func is called, it should be called with a non-nil machine. a nil machine is unexpected and a sign of some problem,so should be raised as an error. If there was no machine, it should have been handled outside of this func and shuold not have called this.